boto3/s3/transfer.py

```diff
@@ -0,0 +1,448 @@
+        self._amount_read = 0
+        self._callback = callback
+
+    @classmethod
+    def from_filename(cls, filename, start_byte, chunk_size, callback=None):
+        """Convenience factory function to create from a filename."""
+        f = open(filename, 'rb')
+        file_size = os.fstat(f.fileno()).st_size
+        return cls(f, start_byte, chunk_size, file_size, callback)
+
+    def _calculate_file_size(self, fileobj, requested_size, start_byte,
+                             actual_file_size):
+        max_chunk_size = actual_file_size - start_byte
+        return min(max_chunk_size, requested_size)
+
+    def read(self, amount=None):
+        if amount is None:
+            amount_to_read = self._size - self._amount_read
+        else:
+            amount_to_read = min(self._size - self._amount_read, amount)
+        data = self._fileobj.read(amount_to_read)
+        self._amount_read += len(data)
+        if self._callback is not None:
+            self._callback(len(data))
+        return data
+
+    def seek(self, where):
+        self._fileobj.seek(self._start_byte + where)
+        self._amount_read = where
+
+    def close(self):
+        self._fileobj.close()
+
+    def tell(self):
+        return self._amount_read
+
+    def __len__(self):
+        # __len__ is defined because requests will try to determine the length
+        # of the stream to set a content length.  In the normal case
+        # of the file it will just stat the file, but we need to change that
+        # behavior.  By providing a __len__, requests will use that instead
+        # of stat'ing the file.
+        return self._size
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+    def __iter__(self):
+        # This is a workaround for http://bugs.python.org/issue17575
+        # Basically httplib will try to iterate over the contents, even
+        # if its a file like object.  This wasn't noticed because we've
+        # already exhausted the stream so iterating over the file immediately
+        # steps, which is what we're simulating here.
+        return iter([])
+
+
+class StreamReaderProgress(object):
+    """Wrapper for a read only stream that adds progress callbacks."""
+    def __init__(self, stream, callback=None):
+        self._stream = stream
+        self._callback = callback
+
+    def read(self, *args, **kwargs):
+        value = self._stream.read(*args, **kwargs)
+        if self._callback is not None:
+            self._callback(len(value))
+        return value
+
+
+class ThreadSafeWriter(object):
+    def __init__(self, write_stream):
+        self._write_stream = write_stream
+        self._lock = threading.Lock()
+
+    def pwrite(self, data, offset):
+        with self._lock:
+            self._write_stream.seek(offset)
+            self._write_stream.write(data)
+
+    def close(self):
+        return self._write_stream.close()
+
+
+class OSUtils(object):
```

Comment:
It's not really clear to me what the purpose of this class of a few one-liners is. I'm concerned that this is an extra layer of abstraction just for the sake of adding another abstraction...